_posts/2016-06-27-bits-of-shapeless-2.md

```diff
@@ -0,0 +1,519 @@
+//                                              ^
+```
+
+To solve this problem there is the concept of a **Coproduct**, the dual concept of a product.
+As oposed to the product, a coproduct represents one element that belong to just one of a set of types. It is like scale's Either extended to an arbitrary number of alternatives. In Shapeless:
+
+```scala
+type stringOrIntCP = String :+: Int :+: CNil
+// defined type alias stringOrIntCP
+```
+
+This type represents something that can be a string or an integer.
+
+Let us see what the coproduct definition looks like in shapeless:
+
+```scala
+sealed trait Coproduct extends Product with Serializable
+
+sealed trait :+:[+H, +T <: Coproduct] extends Coproduct {
+  // Non-recursive fold (like Either#fold)
+  def eliminate[A](l: H => A, r: T => A): A
+}
+
+final case class Inl[+H, +T <: Coproduct](head : H) extends :+:[H, T] {
+  override def eliminate[A](l: H => A, r: T => A) = l(head)
+}
+
+final case class Inr[+H, +T <: Coproduct](tail : T) extends :+:[H, T] {
+  override def eliminate[A](l: H => A, r: T => A) = r(tail)
+}
+
+sealed trait CNil extends Coproduct {
+  def impossible: Nothing
+}
+```
+
+It's a bit more involved than `HList`.
+
+* As analogous to HList's `HNil`, `CNil` marks the end of the type definition, and is used to stop recursion and to possibilitate proper type construction.
+*`:+:` is analogous to Hlist's '::` type constructor, however it can be either an `Inl`or `Inr`.
+* `At runtime if we have and instance of `Inl`we Know we are handling an instance of the "head" of the coproduct.
+* Else if we come across an `Inr`we've got the "tail", which can have either an `Inl` or another nested `Inr` inside. 
+* It is never supposed to reach `CNil`.
+
+The derivation example bellow will make this all much clearer.
+
+Shapeless `LabelledGeneric` is capable of generating a coproduct representation of a sealed trait hierarchy, being the possible values all the known children, tagged with its  class's name -- a labelled coproduct.
+
+```scala
+LabelledGeneric[Shape]
+// res41: shapeless.LabelledGeneric[Shape]{type Repr = shapeless.:+:[Square with shapeless.labelled.KeyTag[Symbol with shapeless.tag.Tagged[String("Square")],Square],shapeless.:+:[Rectangle with shapeless.labelled.KeyTag[Symbol with shapeless.tag.Tagged[String("Rectangle")],Rectangle],shapeless.:+:[Circle with shapeless.labelled.KeyTag[Symbol with shapeless.tag.Tagged[String("Circle")],Circle],shapeless.CNil]]]} = shapeless.LabelledGeneric$$anon$2@503c26b
+```
+
+Note the type of the expression result; all shapes are there, properly tagged.
+
+To make our generic shape json-renderable (and other sealed trait hierarchies) let us tell the compiler how to render `Coproduct`instances into Json.
+
+Although `CNil` should never be reached, we need an implicit `JsonWrites[CNil]` in ope, so the compiler can properly infer the write implicits (stop the recursion):
```

Comment:
in ope? é suposto ser só ope? é um diminutivo de alguma coisa que eu não estou bem a ver ou está errado?